aten/src/ATen/native/cuda/ForeachFunctors.cuh

```diff
@@ -121,6 +121,134 @@ struct AddScalarFunctor {
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                            r_y[ii] = y[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size)
+                            x[i] = r_x[ii];
+                    }
+                }
+            }
+        }
+};
+
+template<typename T>
+struct AddListFunctor {
+    __device__ void operator() (
+        int chunk_size,
+        TensorListMetadata<3>& tl) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
+
+            T* x = (T*)tl.addresses[0][tensor_loc];
+            x += chunk_idx * chunk_size;
+
+            T* y = (T*)tl.addresses[1][tensor_loc];
+            y += chunk_idx * chunk_size;
+
+            T* out = (T*)tl.addresses[2][tensor_loc];
+            out += chunk_idx * chunk_size;
+
+            n -= chunk_idx * chunk_size;
+
+            T r_x[kILP];
+            T r_y[kILP];
+            T r_out[kILP];
+
+            // to make things simple, we put aligned case in a different code path
+            if(n % kILP == 0 && chunk_size % kILP == 0 && is_aligned(x) && is_aligned(y) && is_aligned(out)) {
+                for(int i_start = threadIdx.x; i_start * kILP < n && i_start * kILP < chunk_size; i_start += blockDim.x) {
+                    // load
+                    load_store(r_x, x, 0 , i_start);
+                    load_store(r_y, y, 0 , i_start);
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);
+                    }
+                    // store
+                    load_store(out, r_out, i_start , 0);
+                }
+            }
+            else {
+                // Non-divergent exit condition for __syncthreads, not necessary here
+                for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = 0;
+                        r_y[ii] = 0;
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                            r_y[ii] = y[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);
```

Comment:
It seems like the important part of the Functor is `r_out[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);` and that the other logic is pretty standard between Functors (from me comparing AddListFunctor and AddListFunctor_). Do we ever want functors that don't have similar logic? If not, it might be worth trying to abstract away the rest of the code via templates or macros so that someone implementing a new Functor doesn't have to copy & paste code that is considered to be boilerplate.

Reply:
(to clarify, no action necessary, but I am curious about if we can clean this up in the future)

Reply:
I'm also not sure what the functors will look like in the end state so I agree it makes sense to refactor later